From 8077dacf4cc7c5ee2713afaa5a4b5d9832403025 Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Wed, 26 Jan 2022 09:39:22 -0500
Subject: [PATCH 27/83] ima: Pass template name parameter to
 ima_template_desc_current()

Add a name parameter to the ima_template_desc_current() where callers can
choose the template to use. If callers pass NULL then the default is
automatically chosen.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 security/integrity/ima/ima.h          |  3 ++-
 security/integrity/ima/ima_api.c      |  2 +-
 security/integrity/ima/ima_main.c     |  3 ++-
 security/integrity/ima/ima_policy.c   |  6 +++---
 security/integrity/ima/ima_template.c | 10 +++++++---
 5 files changed, 15 insertions(+), 9 deletions(-)

diff --git a/security/integrity/ima/ima.h b/security/integrity/ima/ima.h
index 76f91c28ef39..9ee4e8aa43b9 100644
--- a/security/integrity/ima/ima.h
+++ b/security/integrity/ima/ima.h
@@ -250,7 +250,8 @@ void ima_print_digest(struct seq_file *m, u8 *digest, u32 size);
 int template_desc_init_fields(const char *template_fmt,
 			      const struct ima_template_field ***fields,
 			      int *num_fields);
-struct ima_template_desc *ima_template_desc_current(struct ima_namespace *ns);
+struct ima_template_desc *ima_template_desc_current(struct ima_namespace *ns,
+						    const char *name);
 struct ima_template_desc *ima_template_desc_buf(void);
 struct ima_template_desc *lookup_template_desc(const char *name);
 bool ima_template_has_modsig(const struct ima_template_desc *ima_template);
diff --git a/security/integrity/ima/ima_api.c b/security/integrity/ima/ima_api.c
index 069a565c767f..a370d0cbf217 100644
--- a/security/integrity/ima/ima_api.c
+++ b/security/integrity/ima/ima_api.c
@@ -48,7 +48,7 @@ int ima_alloc_init_template(struct ima_namespace *ns,
 	if (desc)
 		template_desc = desc;
 	else
-		template_desc = ima_template_desc_current(ns);
+		template_desc = ima_template_desc_current(ns, NULL);
 
 	*entry = kzalloc(struct_size(*entry, template_data,
 				     template_desc->num_fields),
diff --git a/security/integrity/ima/ima_main.c b/security/integrity/ima/ima_main.c
index 919077dc703b..75944a105597 100644
--- a/security/integrity/ima/ima_main.c
+++ b/security/integrity/ima/ima_main.c
@@ -38,7 +38,8 @@ int ima_appraise;
 static int __init hash_setup(struct ima_config *ic, char *str)
 {
 	struct ima_namespace *ns = &init_ima_ns;
-	struct ima_template_desc *template_desc = ima_template_desc_current(ns);
+	struct ima_template_desc *template_desc =
+				ima_template_desc_current(ns, NULL);
 	int i;
 
 	if (ic->hash_setup_done)
diff --git a/security/integrity/ima/ima_policy.c b/security/integrity/ima/ima_policy.c
index 4d6cdde56db3..8cf838daa077 100644
--- a/security/integrity/ima/ima_policy.c
+++ b/security/integrity/ima/ima_policy.c
@@ -713,7 +713,7 @@ int ima_match_policy(struct ima_namespace *ns,
 	struct list_head *ima_rules_tmp;
 
 	if (template_desc && !*template_desc)
-		*template_desc = ima_template_desc_current(ns);
+		*template_desc = ima_template_desc_current(ns, NULL);
 
 	ima_lazy_lsm_update_rules(ns);
 
@@ -1919,7 +1919,7 @@ static int ima_parse_rule(struct user_namespace *user_ns,
 
 	if (!result && entry->flags & IMA_MODSIG_ALLOWED) {
 		template_desc = entry->template ? entry->template :
-						  ima_template_desc_current(ns);
+					ima_template_desc_current(ns, NULL);
 		check_template_modsig(template_desc);
 	}
 
@@ -1927,7 +1927,7 @@ static int ima_parse_rule(struct user_namespace *user_ns,
 	if (!result && entry->action == MEASURE &&
 	    entry->flags & IMA_VERITY_REQUIRED) {
 		template_desc = entry->template ? entry->template :
-						  ima_template_desc_current(ns);
+						  ima_template_desc_current(ns, NULL);
 		check_template_field(template_desc, "d-ngv2",
 				     "verity rules should include d-ngv2");
 	}
diff --git a/security/integrity/ima/ima_template.c b/security/integrity/ima/ima_template.c
index 300a4f7cf816..d81b9ac0b0c9 100644
--- a/security/integrity/ima/ima_template.c
+++ b/security/integrity/ima/ima_template.c
@@ -276,12 +276,15 @@ void ima_init_template_list(void)
 	spin_unlock(&template_list);
 }
 
-struct ima_template_desc *ima_template_desc_current(struct ima_namespace *ns)
+struct ima_template_desc *ima_template_desc_current(struct ima_namespace *ns,
+						    const char *name)
 {
+	const char *tn = name ? name : CONFIG_IMA_DEFAULT_TEMPLATE;
+
 	if (!ns->ima_template) {
 		ima_init_template_list();
 		ns->ima_template =
-		    lookup_template_desc(CONFIG_IMA_DEFAULT_TEMPLATE);
+		    lookup_template_desc(tn);
 	}
 	return ns->ima_template;
 }
@@ -297,7 +300,8 @@ struct ima_template_desc *ima_template_desc_buf(void)
 
 int ima_init_template(struct ima_namespace *ns)
 {
-	struct ima_template_desc *template = ima_template_desc_current(ns);
+	struct ima_template_desc *template =
+			ima_template_desc_current(ns, NULL);
 	int result;
 
 	result = template_desc_init_fields(template->fmt,
-- 
2.37.3

